x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheDirectory.java

```diff
@@ -0,0 +1,156 @@
+ */
+package org.elasticsearch.xpack.searchablesnapshots.cache;
+
+import org.apache.lucene.store.BufferedIndexInput;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.FilterDirectory;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.lease.Releasable;
+import org.elasticsearch.common.lease.Releasables;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * {@link CacheDirectory} uses a {@link CacheService} to cache Lucene files provided by another {@link Directory}.
+ */
+public class CacheDirectory extends FilterDirectory {
+
+    private final CacheService cacheService;
+    private final Path cacheDir;
+
+    public CacheDirectory(Directory in, CacheService cacheService, Path cacheDir) throws IOException {
+        super(in);
+        this.cacheService = Objects.requireNonNull(cacheService);
+        this.cacheDir = Files.createDirectories(cacheDir);
+    }
+
+    public void close() throws IOException {
+        super.close();
+        // Ideally we could let the cache evict/remove cached files by itself after the
+        // directory has been closed.
+        cacheService.removeFromCache(key -> key.startsWith(cacheDir.toString()));
+    }
+
+    @Override
+    public IndexInput openInput(final String name, final IOContext context) throws IOException {
+        ensureOpen();
+        return new CacheBufferedIndexInput(name, fileLength(name), context);
+    }
+
+    public class CacheBufferedIndexInput extends BufferedIndexInput {
+
+        private final String fileName;
+        private final long fileLength;
+        private final Path file;
+        private final IOContext ioContext;
+        private final long offset;
+        private final long end;
+
+        /** a releasable provided by the CacheService that must be released when this IndexInput is done with the cached file **/
+        private @Nullable Releasable releasable;
+        private List<IndexInput> clones;
+        private boolean closed;
+
+        CacheBufferedIndexInput(String fileName, long fileLength, IOContext ioContext) {
+            this(fileName, fileLength, ioContext, "CachedBufferedIndexInput(" + fileName + ")", 0L, fileLength);
+        }
+
+        private CacheBufferedIndexInput(String fileName, long fileLength, IOContext ioContext, String desc, long offset, long length) {
+            super(desc, ioContext);
+            this.fileName = fileName;
+            this.fileLength = fileLength;
+            this.file = cacheDir.resolve(fileName);
+            this.ioContext = ioContext;
+            this.offset = offset;
+            this.end = offset + length;
+            this.clones = new ArrayList<>();
```

Comment:
I think we clone and/or slice from different threads so we should make this threadsafe:

```suggestion
            this.clones = new CopyOnWriteArrayList<>();
```